#505 inbound expiry date edit

joshxg opened the pull request:
Fixes #505 

- Added an expiry date cell to the inbound edit view
![image](https://user-images.githubusercontent.com/35858975/147715556-155fcd72-c97d-4d83-998e-5a8d527fcfff.png)
- Went a little offroad and added some simple keybindings (simple as in, I'm sure there are better ways to do this) for switching between tabs to get some feedback on. I spent a bit trying to figure out what the best key combo was, what the best way to show the keybindings (tooltip for example) and just landed on this as possibly? the worst example (you can't type '!' '@' '#' for example) which I am sure will result in a lot of feedback? 😁

joshxg commented on packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEdit.tsx:
```diff
@@ -208,9 +232,21 @@ export const InboundLineEdit: FC<InboundLineEditProps> = ({
+                      value={Tabs.Location}
+                      label={`${t('label.location')} (⇧+3)`}
+                      tabIndex={-1}
+                    />
```
The main motivation for doing this now, was that the tab order is a little strange when there are multiple rows as the first two columns are a different table.

If we add a 'focus' to the first cell of the 'tabs table' when we change tab, it might be quite nice? Not really sure.. but the alternative is to bring the columns. back into the table which means we have to manually make sure that the widths and stuff are aligned and it's a bit tedious 😁

mark-prins replied:
oh. heh. I commented about exactly that. yeah - a focus would be nice, I think.

joshxg commented on packages/invoices/src/InboundShipment/DetailView/modals/InboundLineEdit/InboundLineEditPanel.tsx:
```diff
@@ -37,16 +38,22 @@ interface InboundLineEditPanel {
   updateDraftLine: (patch: Partial<DraftInboundLine> & { id: string }) => void;
 }
 
+const expiryInputColumn = getExpiryDateInputColumn<DraftInboundLine>();
```
Bit of a gotcha here is that whenever the `updateDraftLine` function changes (When an update happens), we need to get new columns, which means if this is not a stable reference to an object, means the `Cell` will be re-mounted, rather than re-rendered which means it will lose all internal state

joshxg reviewed (commented):
Need to handle validation also. At the moment just a border when there's an error - would like to display feedback but not sure how. Wondering if we should ask Manu or we can just make something up.

I'm thinking because you could have an error on any tab, having a ❗ icon on the far left when the row is in an error state, which you can hover to see what the error is?

mark-prins commented:
general comment re the keybindings - awesome!!
agree, that there might be comments about #@! - but what I like is that SHIFT is fairly transferrable across windows & mac, if we used option or command symbols, that would mean nothing to our current users. 

It would be good to focus the first editable cell on the new tab, if that is at all possible(!) when using keys, I can switch to the location tab easily, but then I have to TAB a whole lot in order to use the drop down. or use the mouse..

mark-prins commented on packages/common/src/utils/formatters.ts:
```diff
@@ -1,8 +1,12 @@
 
-export const formatNaiveDate = (date: Date): string => {
-  return formatDate(date).replace(/\//g, '-');
+export const formatNaiveDate = (date: Date): string | null => {
```
thanks - had an issue with that in the other PR 😄

mark-prins reviewed (approved):
👍  that is looking super slick! 

couple of :nits
- the expiry date formatting in the detail view column: in mSupply this is usually MM/yyyy and currently is d/M/yyyy. my understanding is that the day isn't super useful; I wonder about just having month & year?
- if the expiry is null from the server then today's date shows as the expiry in the edit modal, which made me think that a date was set for the expiry.. but it's just the date picker defaulting. can we have a blank or something?
- seems the API isn't saving the expiry, I had to update it manually to check things

joshxg commented:
Ah yes - have changed the column values.

For not saving, I'm saving against the demo server OK, were you doing something different this? 
https://user-images.githubusercontent.com/35858975/147724234-876583b2-5f2e-4e1e-ac3b-f41ae4d46f51.mp4


....  (I typed that and now an hour later....)

Aaaand now strangely I'm getting errors when I edit records through the API both on the demo server and a newly initialised instance locally. I'm not sure if this is a me issue or if there is something wrong. 

![image](https://user-images.githubusercontent.com/35858975/147733551-94bd37f0-89a7-4871-902b-77d588bfbc9b.png)



Ugh. Going to leave that to the side for the minute..


I've updated the column. Also, I've added a cheap hacky way to focus when changing tabs, which is annoying for the batch/expiry - I think doing a better/deliberate/specific implementation would take a bit of work but might be similar in idea to https://github.com/openmsupply/openmsupply-client/issues/500 where we could create a ref map for the columns... but hopefully this'll give an idea for now

mark-prins commented:
looks great - liking the autofocus change 😁  see what you mean about the refs now!
Having a further look into the API not saving the date - could just be a data thing, so am doing some more sniffing around while writing up the issue for the API team.
